docs/zq/language.md

```diff
@@ -1,341 +1,2091 @@
+[defined below](#expressions).
+
+### 5.1 First-class Types
+
+Like the Zed data model, the Zed language has first class types:
+any Zed type may be used as a value.
+
+The primitive types are listed in the
+[data model specification](../formats/zed.md#1-primitive-types)
+and have the same syntax in the Zed language.  
+Complex types also follow the ZSON syntax.  Note that the type of a type value
+is simply `type`.
+
+As in ZSON, _when types are used as values_, e.g., in a Zed expression,
+they must be referenced within angle brackets.  That is, the integer type
+`int64` is expressed as a type value using the syntax `<int64>`.
+
+Complex types in the Zed language follow the ZSON syntax as well.  
+Here are a few examples:
+```
+* a simple record type - `{x:int64,y:int64}
+* an array of integers - `[int64]`
+* a set of strings - `|[string]|`
+* a map of strings keys to integer values - `{[string,int64]}`
+* a union of string and integer  - `(string,int64)`
+```
+Complex types may be composed, as in `[({s:string},{x:int64})]` which is
+an array of type union of two types of records.
+
+The [typeof function](functions/typeof.md) returns a value's type as
+a value, e.g., `typeof(1)` is `<int64>` and `typeof(<int64>)` is `<type>`.
+
+First-class types are quite powerful because types can
+serve as group-by keys or used in "data shaping" logic.
+A common query for data introspection is to perform some search query
+slicing and dicing some exploratory data then counting the shapes of
+each type of data as follows:
+```
+search ... | count() by typeof(this)
+```
+For example,
+```mdtest-command
+echo '1 2 "foo" 10.0.0.1 <string>' | zq -z 'count() by typeof(this) | sort this' -
+```
+produces
+```mdtest-output
+{typeof:<int64>,count:2(uint64)}
+{typeof:<string>,count:1(uint64)}
+{typeof:<ip>,count:1(uint64)}
+{typeof:<type>,count:1(uint64)}
+```
+When running such a query over complex, semi-structured data, the results can
+be quite illuminating and can inform the design of "data shaping" Zed queries
+to transform raw, messy data into clean data for downstream tooling.
+
+Note the somewhat subtle difference between a record value with a field `t` of
+type `type` whose value is type `string`
+```
+{t:<string>}
+```
+and a record type used as a value
+```
+<{t:string}>
+```
+
+### 5.2 Named Types
+
+As in any modern programming language, types can be named and the type names
+persist into the data model and thus into the serialized input and output.
+
+Types are defined with a type statement of the form
+```
+type <name> = <type>
+```
+where `<name>` is typically an identifier (though a type name can have any
+characters in it so `<name>` can also be a quoted string), and `<type>` is
+any Zed type.  For example,
+```
+type port=uint16
+```
+defines a type named "port" for type `uint16`.
+
+Type statements (along with const statements) must appear at the beginning
+of a query or the beginning of a [parenthesized scope](#scopes).
```

Comment:
This section seems to repeat material in section 4.

Reply:
Indeed, will rework.